crates/red_knot_python_semantic/src/types/infer.rs

```diff
@@ -3882,15 +3874,18 @@ impl<'db> TypeInferenceBuilder<'db> {
 
                 let value_ty = self.infer_expression(value);
 
-                if value_ty
-                    .into_class_literal()
-                    .is_some_and(|ClassLiteralType { class }| {
-                        class.is_known(self.db, KnownClass::Tuple)
-                    })
-                {
-                    self.infer_tuple_type_expression(slice)
-                } else {
-                    self.infer_subscript_type_expression(subscript, value_ty)
+                match value_ty.into_class_literal() {
+                    Some(ClassLiteralType { class })
+                        if class.is_known(self.db, KnownClass::Tuple) =>
+                    {
+                        self.infer_tuple_type_expression(slice)
+                    }
+                    Some(ClassLiteralType { class })
+                        if class.is_known(self.db, KnownClass::Type) =>
+                    {
+                        self.infer_subclass_of_type_expression(slice)
+                    }
+                    _ => self.infer_subscript_type_expression(subscript, value_ty),
                 }
```

Comment:
```suggestion
                match value_ty
                    .into_class_literal()
                    .and_then(|class_ty| class_ty.class.known(self.db))
                {
                    Some(KnownClass::Tuple) => self.infer_tuple_type_expression(slice),
                    Some(KnownClass::Type) => self.infer_subclass_of_type_expression(slice),
                    _ => self.infer_subscript_type_expression(subscript, value_ty),
                }
```

Reply:
That makes sense. What about a nested `match`, in that case? It feels slightly cleaner (and possibly more efficient?) than all these `if class.is_known()` checks in each branch:

```suggestion
                match value_ty {
                    Type::ClassLiteral(class_literal_ty) => {
                        match class_literal_ty.class.known(self.db) {
                            Some(KnownClass::Tuple) => self.infer_tuple_type_expression(slice),
                            Some(KnownClass::Type) => self.infer_subclass_of_type_expression(slice),
                            _ => self.infer_subscript_type_expression(subscript, value_ty),
                        }
                    }
                    _ => self.infer_subscript_type_expression(subscript, value_ty),
                }
```

Reply:
Ok. I don't like that it duplicates the fallback branch, but it looks cleaner, I agree.